### PR TITLE
WIP: add postfix container image

### PR DESCRIPTION
**Pull request by Priyankasaggu11929**
ref: https://github.com/SUSE/BCI-dockerfile-generator/issues/1021

Current state: WIP, testing the generation of Dockerfile with GH workflows, and the output container image for whether/not it works

### CODE REVIEW

File: src/bci_build/package/appcontainers.py [hunk 1]
```diff
@@ -271,6 +271,80 @@
...
+            "cyrus-sasl",
+            "cyrus-sasl-plain",
+            "ca-certificates",
+            "ca-certificates-mozilla",
+            "spamassassin-spamc",
+            "spamass-milter",
+            "timezone",
+            "-busybox-man",
+            "-busybox-sendmail",
+            "-busybox-iproute2",
```
Comment (dcermak) [inline src/bci_build/package/appcontainers.py:303]: You do not have to explicitly remove busybox related packages from the container, as you are using the default base image and not busybox. This should also fix the build failures

File: src/bci_build/package/appcontainers.py [hunk 2]
```diff
@@ -271,6 +271,80 @@
...
+        "version_in_uid": False,
+        "replacements_via_service": [
+            Replacement(
+                regex_in_build_description="%%postfix_version%%",
+                package_name="postfix",
+                parse_version="minor",
+            )
+        ],
+        "package_list": [
+            "bash",
```
Comment (dcermak) [inline src/bci_build/package/appcontainers.py:288]: Bash is preinstalled in the default base image
```suggestion
```

File: src/bci_build/package/appcontainers.py [hunk 3]
```diff
@@ -271,6 +271,80 @@
...
+            "spamassassin-spamc",
+            "spamass-milter",
+            "timezone",
+            "-busybox-man",
+            "-busybox-sendmail",
+            "-busybox-iproute2",
+        ],
+        "entrypoint": ["/entrypoint/entrypoint.sh"],
+        "cmd": ["postfix", "start"],
+        "build_recipe_type": BuildType.DOCKER,
```
Comment (dcermak) [inline src/bci_build/package/appcontainers.py:307]: This is the default
```suggestion
```

File: src/bci_build/package/appcontainers.py [hunk 4]
```diff
@@ -271,6 +271,80 @@
...
+            Path(__file__).parent / "postfix" / "entrypoint" / "ldap" / filename
+        ).read_bytes()
+
+POSTFIX_CONTAINERS = [
+    ApplicationStackContainer(
+        name="postfix",
+        package_name="postfix-image",
+        pretty_name="Postfix",
+        custom_description="Postfix container as fast and secure mail server, {based_on_container}.",
+        **_get_postfix_kwargs(os_version),
```
Comment (dcermak) [inline src/bci_build/package/appcontainers.py:342]: Any particular reason why you opted into putting the kwargs creation into a function and not just explicitly listing them here? Using the indirection via a function usually confuses type checkers & linters and I don't see an immediate benefit here

File: src/bci_build/package/appcontainers.py [hunk 5]
```diff
@@ -271,6 +271,80 @@
...
+        _POSTFIX_FILES[filename] = (
+            Path(__file__).parent / "postfix" / "entrypoint" / "ldap" / filename
+        ).read_bytes()
+
+POSTFIX_CONTAINERS = [
+    ApplicationStackContainer(
+        name="postfix",
+        package_name="postfix-image",
+        pretty_name="Postfix",
+        custom_description="Postfix container as fast and secure mail server, {based_on_container}.",
```
Comment (dcermak) [inline src/bci_build/package/appcontainers.py:341]:
```suggestion
        custom_description="Postfix container is fast and secure mail server, {based_on_container}.",
```

**Review by dcermak — CHANGES_REQUESTED** [review]
Looks like a good start, thanks a lot!

Few points:
- It would be great to have a readme
- The added files bear a lot of resemblance to https://github.com/thkukuk/containers-mailserver/tree/master/postfix, could you either add them to https://github.com/SUSE/BCI-dockerfile-generator/blob/main/update-files.sh or convince Thorsten to make BCI upstream?

File: src/bci_build/package/appcontainers.py (hunk 1 above)
Comment (Priyankasaggu11929) [reply src/bci_build/package/appcontainers.py]: addressed

File: src/bci_build/package/appcontainers.py (hunk 4 above)
Comment (Priyankasaggu11929) [reply src/bci_build/package/appcontainers.py]: No specific reason. I was using the nginx containers (for reference) & ended up copy/pasting.

Listed them explicitly now.

**Comment by Priyankasaggu11929** [comment]
> It would be great to have a readme

Yes, will add a README.

> The added files bear a lot of resemblance to https://github.com/thkukuk/containers-mailserver/tree/master/postfix, could you either add them to https://github.com/SUSE/BCI-dockerfile-generator/blob/main/update-files.sh or convince Thorsten to make BCI upstream?

For the PR, I've added all postfix container files to the update-files.sh. Will reach out to Thorsten.

File: update-files.sh [hunk 6]
```diff
@@ -1,6 +1,7 @@
 #!/bin/bash -e
 
 mkdir -p src/bci_build/package/{postgres,nginx,grafana,mariadb}
+mkdir -p src/bci_build/package/postfix/entrypoint/ldap
```
Comment (dcermak) [inline update-files.sh:4]: We started the one-liner rabbit hole, so let's die on that hill :stuck_out_tongue_winking_eye:  
```suggestion
mkdir -p src/bci_build/package/{postgres,nginx,grafana,mariadb,postfix/entrypoint/ldap}
```

Comment (Priyankasaggu11929) [reply update-files.sh]: addressed in the latest commit refresh.

**Comment by Priyankasaggu11929** [comment]
The SP5/6 builds are complaining for missing dependency packages, is there a requirement to add those packages to these code streams?

Or build the container for few os versions only?

**Comment by dcermak** [comment]
> The SP5/6 builds are complaining for missing dependency packages, is there a requirement to add those packages to these code streams?
> 
> Or build the container for few os versions only?

We do not need to have the postfix container in SLES, it is only required for SLCC, so it would be ok to skip it for SLES. However, we will most likely have exactly the same issue with the missing packages in SLCC.

Do you see an easy way how to build the container without the missing packages for SLES/SLCC (assume for now, that they have the same package set)?

**Comment by Priyankasaggu11929** [comment]
>Do you see an easy way how to build the container without the missing packages for SLES/SLCC (assume for now, that they have the same package set)?

I'll try & get back.

**Comment by Priyankasaggu11929** [comment]
@dcermak, just an update. I'll refresh  the PR with following changes:

- For `os_version: {Tumbleweed}`
    Since TW repo has all dependency packages (including - `spamassassin-spamc`, `spamass-milter`, `mandoc`).
    I will add a conditional to build it as it is (current state of PR)

- For man pages in case of  `os_version: {SLES, SLCC}`:
    - I'll add the `custom_end` field, to create a placeholder file, with a link to online postfix man pages - https://www.postfix.org/postfix-manuals.html.
    And, the echo the same whenever the container starts.
    - And then remove `mandoc` package dependency.

- For spamassasin packages in  `os_version: {SLES/SLCC}`:
     - I'll make a copy of existing `entrypoint.sh`. Remove:
        - `setup_spamassassin()`, `stop_spamassassin()` functions
        - any conditionals triggered by setting env `SPAMASSASSIN_HOST`
    - and then remove rest of the missing packages: `pamassassin-spamc, spamass-milter` 

----

For the README, the existing doc [here](https://github.com/thkukuk/containers-mailserver/blob/master/postfix/README.md) is very thoroughly explaining all the ENV for different postfix configs / server setups, and it's fairly easy to follow.

I'm thinking of adding it as it as, just a different copy for TW & SLCC/SLES each:
    - For, TW, replace the opensuse registry references
    - For SLCC/SLES, above + remove the spamassasin bits, as well.

File: src/bci_build/package/appcontainers.py [hunk 7]
```diff
@@ -271,6 +271,101 @@
...
+                    "spamass-milter",
+                )
+                if os_version == OsVersion.TUMBLEWEED
+                else ()
+            )
+        ],
+        entrypoint=["/entrypoint/entrypoint.sh && cat /entrypoint/postfix-manpages.md"],
+        cmd=["postfix", "start"],
+        extra_files=_POSTFIX_FILES,
+        support_level=SupportLevel.L3,  # TODO: confirm support level
```
Comment (Priyankasaggu11929) [inline src/bci_build/package/appcontainers.py:356]: @dcermak, PR is ready for another round of review.

Summary of recent changes - https://github.com/SUSE/BCI-dockerfile-generator/pull/1046#issuecomment-2077268871

Ques - what is the support level for this container image?

File: src/bci_build/package/appcontainers.py [hunk 8]
```diff
@@ -271,6 +271,101 @@
...
+            Replacement(
+                regex_in_build_description="%%postfix_version%%",
+                package_name="postfix",
+                parse_version="minor",
+            )
+        ],
+        package_list=[
+            name
+            for name in (
+                "ca-certificates",
```
Comment (dirkmueller) [inline src/bci_build/package/appcontainers.py:332]: this is already included in the base container iirc

Comment (Priyankasaggu11929) [reply src/bci_build/package/appcontainers.py]: Removed `ca-certificates` from the package list. Thanks

**Comment by Priyankasaggu11929** [comment]
> nothing provides product(SL-Micro) needed by registries-conf-sl-micro

@dcermak @dirkmueller, Basalt builds are failing with ^.

What is needed to resolve the `product(SL-Micro)` requirement?

**Comment by dirkmueller** [comment]
> What is needed to resolve the `product(SL-Micro)` requirement?

we don't actually want SL-micro inside the container image. so we need to inject a dummy release product.

**Comment by Priyankasaggu11929** [comment]
@dcermak, README changes are up-to-date now with https://github.com/SUSE/BCI-dockerfile-generator/blob/main/README.rst#documentation---readmes

File: update-files.sh [hunk 9]
```diff
@@ -16,4 +17,14 @@ curl -o src/bci_build/package/grafana/run.sh https://raw.githubusercontent.com/g
...
+curl -o src/bci_build/package/postfix/entrypoint/ldap/smtpd_sender_login_maps https://raw.githubusercontent.com/thkukuk/containers-mailserver/master/postfix/ldap/smtpd_sender_login_maps
+curl -o src/bci_build/package/postfix/entrypoint/ldap/virtual_alias_domains https://raw.githubusercontent.com/thkukuk/containers-mailserver/master/postfix/ldap/virtual_alias_domains
+curl -o src/bci_build/package/postfix/entrypoint/ldap/virtual_alias_maps https://raw.githubusercontent.com/thkukuk/containers-mailserver/master/postfix/ldap/virtual_alias_maps
+curl -o src/bci_build/package/postfix/entrypoint/ldap/virtual_gid_maps https://raw.githubusercontent.com/thkukuk/containers-mailserver/master/postfix/ldap/virtual_gid_maps
+curl -o src/bci_build/package/postfix/entrypoint/ldap/virtual_mailbox_maps https://raw.githubusercontent.com/thkukuk/containers-mailserver/master/postfix/ldap/virtual_mailbox_maps
+curl -o src/bci_build/package/postfix/entrypoint/ldap/virtual_uid_maps https://raw.githubusercontent.com/thkukuk/containers-mailserver/master/postfix/ldap/virtual_uid_maps
+>>>>>>> 6200bc41 (add postfix container image)
```
Comment (dcermak) [inline update-files.sh:30]: Merge conflict?
```suggestion
curl -o src/bci_build/package/templates/rmt_helm_chart_readme.j2 https://raw.githubusercontent.com/SUSE/helm-charts/main/rmt-helm/README.md
curl -o src/bci_build/package/postfix/entrypoint/entrypoint.sh https://raw.githubusercontent.com/thkukuk/containers-mailserver/master/postfix/entrypoint.sh
curl -o src/bci_build/package/postfix/entrypoint/ldap/smtpd_sender_login_maps https://raw.githubusercontent.com/thkukuk/containers-mailserver/master/postfix/ldap/smtpd_sender_login_maps
curl -o src/bci_build/package/postfix/entrypoint/ldap/virtual_alias_domains https://raw.githubusercontent.com/thkukuk/containers-mailserver/master/postfix/ldap/virtual_alias_domains
curl -o src/bci_build/package/postfix/entrypoint/ldap/virtual_alias_maps https://raw.githubusercontent.com/thkukuk/containers-mailserver/master/postfix/ldap/virtual_alias_maps
curl -o src/bci_build/package/postfix/entrypoint/ldap/virtual_gid_maps https://raw.githubusercontent.com/thkukuk/containers-mailserver/master/postfix/ldap/virtual_gid_maps
curl -o src/bci_build/package/postfix/entrypoint/ldap/virtual_mailbox_maps https://raw.githubusercontent.com/thkukuk/containers-mailserver/master/postfix/ldap/virtual_mailbox_maps
curl -o src/bci_build/package/postfix/entrypoint/ldap/virtual_uid_maps https://raw.githubusercontent.com/thkukuk/containers-mailserver/master/postfix/ldap/virtual_uid_maps
```

Comment (Priyankasaggu11929) [reply update-files.sh]: fixed in recent commit fresh.

File: src/bci_build/package/appcontainers.py [hunk 10]
```diff
@@ -249,6 +249,93 @@
...
+                (
+                    "mandoc",
+                    "spamassassin-spamc",
+                    "spamass-milter",
+                )
+                if os_version == OsVersion.TUMBLEWEED
+                else ()
+            )
+        ],
+        entrypoint=["/entrypoint/entrypoint.sh && cat /entrypoint/postfix-manpages.md"],
```
Comment (dcermak) [inline src/bci_build/package/appcontainers.py:320]: This does not work, the entrypoint is not passed via a shell but executed directly
```suggestion
        entrypoint=["/entrypoint/entrypoint.sh"],
```

File: src/bci_build/package/appcontainers.py [hunk 11]
```diff
@@ -249,6 +249,93 @@
...
+        support_level=SupportLevel.TECHPREVIEW,
+        exposes_tcp=[25, 465, 587],
+        volumes=["/var/spool/postfix", "/var/spool/vmail", "/etc/pki"],
+        custom_end=f"""{DOCKERFILE_RUN} mkdir -p /entrypoint/ldap
+COPY entrypoint.sh postfix-manpages.md /entrypoint/
+{DOCKERFILE_RUN} chmod +x /entrypoint/entrypoint.sh
+COPY smtpd_sender_login_maps virtual_alias_domains virtual_alias_maps virtual_gid_maps virtual_mailbox_maps virtual_uid_maps /entrypoint/ldap/
+
+HEALTHCHECK --interval=5s --timeout=10s --start-period=30s --retries=3 \
+        CMD postfix status
```
Comment (dcermak) [inline src/bci_build/package/appcontainers.py:332]: Does this work for you? When I try the built container, then it never becomes healthy as `postfix status` reports that the service is not running

```suggestion
HEALTHCHECK --interval=5s --timeout=10s --start-period=30s --retries=3 \
        CMD postfix status
```

Comment (Priyankasaggu11929) [reply src/bci_build/package/appcontainers.py:332]: @dcermak, locally, I've tried building the container image with a `--format docker` flag to buildah.
Without the flag, there is no HEALTHCHECK created.

`WARN[0177] HEALTHCHECK is not supported for OCI image format and will be ignored. Must use `docker` format`

---

And with the recent postfix version bump to 3.9.x in Factory, I do need to update the ldap maps, for the below warnings. But the container gets healthy after 1/2 tries.


```
> sudo buildah bud --format docker -t postfix-image .

> sudo podman run --rm -it --name postfix -p 8488:25  -e SERVER_HOSTNAME=localhost postfix-image

> sudo podman inspect postfix --format "{{json .State.Health}}" | jq
{
  "Status": "healthy",
  "FailingStreak": 0,
  "Log": [
    {
      "Start": "2024-05-29T12:06:19.0453439+05:30",
      "End": "2024-05-29T12:06:20.079235802+05:30",
      "ExitCode": 1,
      "Output": "postfix: fatal: bad string length 0 < 1: myhostname = "
    },
    {
      "Start": "2024-05-29T12:06:50.882187668+05:30",
      "End": "2024-05-29T12:06:50.928231067+05:30",
      "ExitCode": 0,
      "Output": "/usr/sbin/postconf: warning: /etc/postfix/main.cf: support for parameter \"smtp_use_tls\" will be removed; instead, specify \"smtp_tls_security_level\"\n/usr/sbin/postconf: warning: /etc/postfix/main.cf: support for parameter \"smtpd_enforce_tls\" will be removed; instead, specify \"smtpd_tls_security_level\"\n/usr/sbin/postconf: warning: /etc/postfix/main.cf: support for parameter \"smtpd_use_tls\" will be removed; instead, specify \"smtpd_tls_security_level\"\n/usr/sbin/postconf: warning: /etc/postfix/main.c"
    }
  ]
}

```

File: src/bci_build/package/appcontainers.py (hunk 10 above)
Comment (Priyankasaggu11929) [reply src/bci_build/package/appcontainers.py:320]: moving the text in `postfix-manpages.md` to `entrypoint.sles.sh` as a echo command, and removing the md file after.

File: src/bci_build/package/appcontainers.py (hunk 11 above)
Comment (dcermak) [reply src/bci_build/package/appcontainers.py:332]: Please try out the image from OBS. The bot created a test build and added registry links in the comments, e.g. here: https://github.com/SUSE/BCI-dockerfile-generator/pull/1046#issuecomment-2134576192

Unfortunately OBS builds images differently than a plain `buildah bud` or `docker build`.

File: src/bci_build/package/appcontainers.py [hunk 12]
```diff
@@ -249,6 +249,93 @@
...
+        replacements_via_service=[
+            Replacement(
+                regex_in_build_description="%%postfix_version%%",
+                package_name="postfix",
+                parse_version="minor",
+            )
+        ],
+        package_list=[
+            nameRefer to Postfix manual pages at https://www.postfix.org/postfix-manuals.html
+
```
Comment (dcermak) [inline src/bci_build/package/appcontainers.py:298]:
```suggestion
            name
```

File: src/bci_build/package/appcontainers.py (hunk 11 above)
Comment (Priyankasaggu11929) [reply src/bci_build/package/appcontainers.py:331]: Oh, I totally missed the registry links for staging images all this time. Thanks for pointing. 

Testing again!

Comment (Priyankasaggu11929) [reply src/bci_build/package/appcontainers.py:331]: @dcermak, after the fix in https://github.com/SUSE/BCI-dockerfile-generator/pull/1046#discussion_r1618291096, the HEALTHCHECK in the built image also works fine for me.

I followed:

```
> sudo podman run --rm -it --name postfix -p 8684:25 -e SERVER_HOSTNAME=localhost registry.opensuse.org/home/defolos/bci/staging/tumbleweed/tumbleweed-1046/containerfile/opensuse/postfix

> podman inspect postfix --format "{{json .State.Health}}" | jq
{
  "Status": "healthy",
  "FailingStreak": 0,
  "Log": [
    {
      "Start": "2024-05-29T21:04:45.362446147+05:30",
      "End": "2024-05-29T21:04:45.418870472+05:30",
      "ExitCode": 0,
      "Output": "/usr/sbin/postconf: warning: /etc/postfix/main.cf: support for parameter \"smtp_use_tls\" will be removed; instead, specify \"smtp_tls_security_level\"\n/usr/sbin/postconf: warning: /etc/postfix/main.cf: support for parameter \"smtp_enforce_tls\" will be removed; instead, specify \"smtp_tls_security_level\"\n/usr/sbin/postconf: warning: /etc/postfix/main.cf: support for parameter \"smtpd_use_tls\" will be removed; instead, specify \"smtpd_tls_security_level\"\n/usr/sbin/postconf: warning: /etc/postfix/main.cf:"
    },
   ...
   ...
   ]
```

File: src/bci_build/package/appcontainers.py [hunk 13]
```diff
@@ -138,6 +138,84 @@ def _envsubst_pkg_name(os_version: OsVersion) -> str:
...
+            )
+        ],
+        entrypoint=["/entrypoint/entrypoint.sh"],
+        cmd=["postfix", "start"],
+        extra_files=_POSTFIX_FILES,
+        support_level=SupportLevel.TECHPREVIEW,
+        exposes_tcp=[25, 465, 587],
+        volumes=["/var/spool/postfix", "/var/spool/vmail", "/etc/pki"],
+        custom_end=f"""{DOCKERFILE_RUN} mkdir -p /entrypoint/ldap
+COPY {"entrypoint.sh" if os_version == OsVersion.TUMBLEWEED else "entrypoint.sles.sh /entrypoint/entrypoint.sh"}
```
Comment (dcermak) [inline src/bci_build/package/appcontainers.py:209]:
```suggestion
COPY {"entrypoint.sh" if os_version == OsVersion.TUMBLEWEED else "entrypoint.sles.sh"} /entrypoint/entrypoint.sh
```

**Comment by Priyankasaggu11929** [comment]
@dcermak @dirkmueller, is this PR in a (early-ish) good enough state to merge?

This will then help verify the smoke testing in the pipeline – https://github.com/SUSE/BCI-tests/pull/487

I'll continue improving the further bits on this container in follow up PRs.

File: src/bci_build/package/appcontainers.py [hunk 14]
```diff
@@ -138,6 +138,85 @@ def _envsubst_pkg_name(os_version: OsVersion) -> str:
...
+            )
+        ],
+        entrypoint=["/entrypoint/entrypoint.sh"],
+        cmd=["postfix", "start"],
+        extra_files=_POSTFIX_FILES,
+        support_level=SupportLevel.TECHPREVIEW,
+        exposes_tcp=[25, 465, 587],
+        volumes=["/var/spool/postfix", "/var/spool/vmail", "/etc/pki"],
+        custom_end=f"""{DOCKERFILE_RUN} mkdir -p /entrypoint/ldap
+COPY {"entrypoint.sh" if os_version == OsVersion.TUMBLEWEED else "entrypoint.sles.sh"} /entrypoint/entrypoint.sh
```
Comment (dirkmueller) [inline src/bci_build/package/appcontainers.py:209]: where is "entrypoint.sles.sh" coming from? it's not coming from `update-files.sh` does it? I would appreciate if we would also have an origin of it. maybe you can generate it from `entrypoint.sh` by applying a patch or running `sed` ?

**Review by dirkmueller — CHANGES_REQUESTED** [review]
LGTM, except for the unclear entrypoint.sles.sh origin (see inline comment). with that clarified it will be :+1:

File: src/bci_build/package/appcontainers.py (hunk 14 above)
Comment (Priyankasaggu11929) [reply src/bci_build/package/appcontainers.py:209]: > I would appreciate if we would also have an origin of it. maybe you can generate it from entrypoint.sh by applying a patch or running sed ?

working on it.

File: update-files.sh [hunk 15]
```diff
@@ -1,23 +1,29 @@
...
+curl -o src/bci_build/package/postfix/entrypoint/ldap/virtual_alias_domains https://raw.githubusercontent.com/thkukuk/containers-mailserver/master/postfix/ldap/virtual_alias_domains
+curl -o src/bci_build/package/postfix/entrypoint/ldap/virtual_alias_maps https://raw.githubusercontent.com/thkukuk/containers-mailserver/master/postfix/ldap/virtual_alias_maps
+curl -o src/bci_build/package/postfix/entrypoint/ldap/virtual_gid_maps https://raw.githubusercontent.com/thkukuk/containers-mailserver/master/postfix/ldap/virtual_gid_maps
+curl -o src/bci_build/package/postfix/entrypoint/ldap/virtual_mailbox_maps https://raw.githubusercontent.com/thkukuk/containers-mailserver/master/postfix/ldap/virtual_mailbox_maps
+curl -o src/bci_build/package/postfix/entrypoint/ldap/virtual_uid_maps https://raw.githubusercontent.com/thkukuk/containers-mailserver/master/postfix/ldap/virtual_uid_maps
+
+patch src/bci_build/package/postfix/entrypoint/entrypoint.sh src/bci_build/package/postfix/entrypoint/sles-entrypoint.patch -o src/bci_build/package/postfix/entrypoint/entrypoint.sles.sh
```
Comment (Priyankasaggu11929) [inline update-files.sh:29]: @dirkmueller,  above should address the review comment in https://github.com/SUSE/BCI-dockerfile-generator/pull/1046#discussion_r1684555993

Please review again! Thank you!

File: src/bci_build/package/appcontainers.py [hunk 16]
```diff
@@ -138,6 +138,85 @@ def _envsubst_pkg_name(os_version: OsVersion) -> str:
...
+{DOCKERFILE_RUN} chmod +x /entrypoint/entrypoint.sh
+COPY smtpd_sender_login_maps virtual_alias_domains virtual_alias_maps virtual_gid_maps virtual_mailbox_maps virtual_uid_maps /entrypoint/ldap/
+HEALTHCHECK --interval=5s --timeout=10s --start-period=30s --retries=3 \
+        CMD postfix status
+""",
+    )
+    for os_version in set(ALL_NONBASE_OS_VERSIONS) | {OsVersion.BASALT}
```
Comment (dcermak) [inline src/bci_build/package/appcontainers.py:216]: Basalt is now gone from main:
```suggestion
    for os_version in set(ALL_NONBASE_OS_VERSIONS) | {OsVersion.SLE16_0}
```

File: src/bci_build/package/postfix/README.md.j2 [hunk 17]
```diff
@@ -0,0 +1,183 @@
...
+The environment variable `VIRTUAL_MBOX=1` will enable support for virtual
+domains and virtual users. It can be used in conjunction with a relay host.
+
+Mails for virtual domains and users are stored in Maildir format in
+`/var/spool/vmail` inside the container. To not loose the mail, this directory
+should be a persistence volume.
+`VMAIL_UID` specifies the UID and GID which owns all files below this
+directory. This makes sure, that the container is using the same UID/GID as
+the Container Host OS for the files and not regular users are owning this
+files and can read and modify them. The default UID/Gid is `5000`.
```
Comment (dcermak) [inline src/bci_build/package/postfix/README.md.j2:96]:
```suggestion
files and can read and modify them. The default UID/GID is `5000`.
```

File: src/bci_build/package/postfix/README.md.j2 [hunk 18]
```diff
@@ -0,0 +1,183 @@
...
+define the virtual domains and the virtual users.
+`VIRTUAL_DOMAINS="example.com example2.com example3.com"` is a whitespace
+seperated list which can contain one or more virtual domains. For every
+domain, an own directory below `/var/spool/vmail` will be created.
+`VIRTUAL_USERS="user1@example.com user@example3.com user2@example.com"`
+Whitespace seperated list of virtual users email addresses.
+
+The example call:
+```sh
+podman run -d --rm --name postfix -p "25:25" -e VIRTUAL_MBOX=1 -e VMAIL_UID=5000 -e VIRTUAL_DOMAINS="example.com example1.com" -e VIRTUAL_USERS="user1@example.com user2@example.com user@example1.com" -e SERVER_HOSTNAME=smtp.example.com -e SMTP_RELAYHOST=relay.example.com -e SMTP_USERNAME=mailer -e SMTP_PASSWORD='XXX' -v "/srv/postfix/vmail:/var/spool/vmail" {{ image.pretty_reference }}
```
Comment (dcermak) [inline src/bci_build/package/postfix/README.md.j2:113]:
```suggestion
podman run -d --rm --name postfix -p "25:25" \
    -e VIRTUAL_MBOX=1 \
    -e VMAIL_UID=5000 \
    -e VIRTUAL_DOMAINS="example.com example1.com" \
    -e VIRTUAL_USERS="user1@example.com user2@example.com user@example1.com" \
    -e SERVER_HOSTNAME=smtp.example.com \
    -e SMTP_RELAYHOST=relay.example.com \
    -e SMTP_USERNAME=mailer \
    -e SMTP_PASSWORD='XXX' \
    -v "/srv/postfix/vmail:/var/spool/vmail" \
    {{ image.pretty_reference }}
```

File: src/bci_build/package/postfix/README.md.j2 [hunk 19]
```diff
@@ -0,0 +1,183 @@
...
+of the virtual users mailbox. To disable it, the file would look like:
+```
+user1@example.com 0
+user2@example.com 0
+user@example1.com 0
+```
+
+The example call:
+```sh
+podman run -d --rm --name postfix -p 25:25 -e VIRTUAL_MBOX=1 -e VMAIL_UID=5000 -e SERVER_HOSTNAME=smtp.example.com -e SMTP_RELAYHOST=relay.example.com -e SMTP_USERNAME=mailer -e SMTP_PASSWORD='XXX' -v "/srv/postfix/vmail:/var/spool/vmail" -v "/srv/postfix/etc/vhosts:/etc/postfix/vhosts:ro" -v "/srv/postfix/etc/vmaps:/etc/postfix/vmaps:ro" -v "/srv/postfix/etc/vquota:/etc/postfix/vquota:ro" {{ image.pretty_reference }}
```
Comment (dcermak) [inline src/bci_build/package/postfix/README.md.j2:164]:
```suggestion
podman run -d --rm --name postfix -p 25:25 \
    -e VIRTUAL_MBOX=1 \
    -e VMAIL_UID=5000 \
    -e SERVER_HOSTNAME=smtp.example.com \
    -e SMTP_RELAYHOST=relay.example.com \
    -e SMTP_USERNAME=mailer \
    -e SMTP_PASSWORD='XXX' \
    -v "/srv/postfix/vmail:/var/spool/vmail:z" \
    -v "/srv/postfix/etc/vhosts:/etc/postfix/vhosts:z,ro" \
    -v "/srv/postfix/etc/vmaps:/etc/postfix/vmaps:z,ro" \
    -v "/srv/postfix/etc/vquota:/etc/postfix/vquota:z,ro" \
    {{ image.pretty_reference }}
```

File: update-files.sh [hunk 20]
```diff
@@ -1,23 +1,29 @@
...
-curl -sf -o src/bci_build/package/nginx/30-tune-worker-processes.sh https://raw.githubusercontent.com/nginxinc/docker-nginx/master/entrypoint/30-tune-worker-processes.sh
+curl -o src/bci_build/package/nginx/docker-entrypoint.sh https://raw.githubusercontent.com/nginxinc/docker-nginx/master/entrypoint/docker-entrypoint.sh
+curl -o src/bci_build/package/nginx/LICENSE https://raw.githubusercontent.com/nginxinc/docker-nginx/master/LICENSE
+curl -o src/bci_build/package/nginx/20-envsubst-on-templates.sh https://raw.githubusercontent.com/nginxinc/docker-nginx/master/entrypoint/20-envsubst-on-templates.sh
+curl -o src/bci_build/package/nginx/30-tune-worker-processes.sh https://raw.githubusercontent.com/nginxinc/docker-nginx/master/entrypoint/30-tune-worker-processes.sh
 
-curl -sf -o src/bci_build/package/grafana/run.sh https://raw.githubusercontent.com/grafana/grafana/main/packaging/docker/run.sh
+curl -o src/bci_build/package/grafana/run.sh https://raw.githubusercontent.com/grafana/grafana/main/packaging/docker/run.sh
```
Comment (dcermak) [inline update-files.sh:14]: please keep the `-sf` here

**Review by dcermak — COMMENTED** [review]
LGTM, but needs a rebase and please take a look at the few comments that I added

**Comment by Priyankasaggu11929** [comment]
> LGTM, but needs a rebase and please take a look at the few comments that I added

@dcermak, have rebased the PR & addressed all recent comments.

**Review by dcermak — APPROVED** [review]

**Review by dirkmueller — APPROVED** [review]

**Pull request merged**